File: src/iac_code/ui/banner.py
"""Welcome banner rendering."""

from __future__ import annotations

import getpass
import shlex
from collections.abc import Iterable
from pathlib import Path
from typing import TYPE_CHECKING

from rich.align import Align
from rich.color import Color
from rich.console import Group
from rich.panel import Panel
from rich.style import Style
from rich.table import Table
from rich.text import Text

from iac_code.i18n import _

if TYPE_CHECKING:
    from rich.console import Console

    from iac_code.services.update_checker import PendingUpdate

# The app icon uses a broad stroke that turns into several parallel dot rows
# when encoded directly as Braille. This terminal-specific trace uses a
# single-dot stroke. Its outer modules are exact quarter-turns of one master,
# so the right corners join cleanly without isolated or protruding dots.
BRAILLE_LOGO_LINES = (
    " ⢀⠴⠒⠒⠒⠂⡠⠒⠒⠒⠦⡀",
    "⢰⠃   ⡠⠊     ⠘⡆",
    "⢸   ⠊ ⢠  ⠑⢄  ⡇",
    "⢀⠑⢄  ⢈⢾⡮⠤  ⠑⢄⠁",
    "⢸  ⠑⢄ ⠘⠈ ⡠   ⡇",
    "⠸⡄     ⡠⠊   ⢠⠇",
    " ⠈⠲⠤⠤⠤⠊⠠⠤⠤⠤⠖⠁",
)

BRAND_GRADIENT = (
    "#45b4ff",
    "#4388ff",
    "#5968f5",
    "#7d63f4",
    "#a969ef",
    "#e778cb",
)

ACCENT = "bright_cyan"


def _logo_color(row: int, column: int) -> Color:
    """Pick the nearest brand-gradient stop for a terminal logo cell."""
    width = max(len(line) for line in BRAILLE_LOGO_LINES)
    extent = len(BRAILLE_LOGO_LINES) + width - 2
    position = (row + column) / extent
    index = round(position * (len(BRAND_GRADIENT) - 1))
    return Color.parse(BRAND_GRADIENT[index])


def _render_pixel_logo() -> Text:
    """Render the compact Braille fallback logo."""
    logo = Text()
    width = max(len(line) for line in BRAILLE_LOGO_LINES)

    for row, line in enumerate(BRAILLE_LOGO_LINES):
        if row:
            logo.append("\n")
        logo.append("   ")
        for column, cell in enumerate(line.ljust(width)):
            if cell != " ":
                logo.append(cell, Style(color=_logo_color(row, column)))
            else:
                logo.append(" ")

    return logo


def _render_image_placeholder() -> Text:
    """Reserve the same terminal cells used by the Braille fallback."""
    width = max(len(line) for line in BRAILLE_LOGO_LINES) + 3
    return Text("\n".join(" " * width for _ in BRAILLE_LOGO_LINES))


def _format_update_command(command: Iterable[str]) -> str:
    return shlex.join(tuple(command))


def render_update_prompt_header(update: PendingUpdate) -> Group:
    """Render update information above the interactive update prompt."""
    command_text = _format_update_command(update.update_command)
    items = [
        Text(_("Update available! {} -> {}").format(update.current_version, update.version), style="bold bright_cyan"),
        Text("{}: {}".format(_("Update command"), command_text), style="bold"),
    ]
    if update.release_notes_url:
        items.append(Text("{}: {}".format(_("Release notes"), update.release_notes_url), style="dim"))
    return Group(*items)


def render_update_notice(update: PendingUpdate) -> Panel:
    """Render a notice for an update the user previously skipped."""
    command_text = _format_update_command(update.update_command)
    items = [
        Text(_("Update available! {} -> {}").format(update.current_version, update.version), style="bold bright_cyan"),
        Text(_("Run {} to update.").format(command_text)),
    ]
    if update.release_notes_url:
        items.append(Text("{}: {}".format(_("Release notes"), update.release_notes_url), style="dim"))
    return Panel(Group(*items), border_style=ACCENT, expand=True)


def _get_provider_display() -> str:
    """Get the active provider display name from settings."""
    try:
        from iac_code.config import PARTNER_SOURCES, get_active_provider_key, get_llm_source, get_provider_config
        from iac_code.i18n import _
        from iac_code.providers.registry import PROVIDER_REGISTRY

        key = get_active_provider_key()
        if not key:
            llm_source = get_llm_source()
            for ps in PARTNER_SOURCES:
                if ps.key == llm_source:
                    real_provider = ps.get_provider_display()
                    if real_provider:
                        return "{} / {}".format(ps.display_name, real_provider)
                    return ps.display_name
            return ""
        desc = PROVIDER_REGISTRY.get(key)
        if desc:
            return _(desc.display_name)
        name = get_provider_config(key).get("name", "")
        return name
    except Exception:
        return ""


def render_welcome_banner(
    model: str,
    cwd: str,
    session_id: str | None = None,
    session_name: str | None = None,
    *,
    _image_placeholder: bool = False,
) -> Panel:
    """Produce a Rich Panel for the welcome banner."""
    # Username
    try:
        username = getpass.getuser()
        username = username[0].upper() + username[1:] if username else "User"
    except Exception:
        username = "User"

    # Logo
    logo = _render_image_placeholder() if _image_placeholder else _render_pixel_logo()

    # Description (centered vertically beside the logo)
    desc_text = Text(_("Your AI-powered Infrastructure as Code assistant"), style="italic white")

    # Use a table for side-by-side layout with vertical centering
    logo_table = Table(show_header=False, show_edge=False, box=None, padding=0, expand=True)
    logo_table.add_column(ratio=1)
    logo_table.add_column(ratio=2)
    logo_table.add_row(logo, Align(desc_text, align="center", vertical="middle"))

    # Shorten cwd
    cwd_path = Path(cwd).resolve()
    try:
        cwd_display = "~/" + cwd_path.relative_to(Path.home()).as_posix()
    except ValueError:
        cwd_display = str(cwd_path)

    # Provider / model display
    provider_name = _get_provider_display()
    if provider_name and model:
        model_display = f"{provider_name} / {model}"
    else:
        model_display = model

    from iac_code import __version__

    session_display: Text
    if session_name and session_id:
        session_display = Text("  {}: {} ({})".format(_("Session"), session_name, session_id), style="dim")
    elif session_id:
        session_display = Text("  {}: {}".format(_("Session"), session_id), style="dim")
    else:
        session_display = Text()

    items = [
        Text(),
        Text("  {} {}!".format(_("Welcome back"), username), style="bold"),
        Text(),
        logo_table,
        Text(),
        Text(f"  iac-code v{__version__}", style="dim"),
        Text(f"  {model_display}", style="dim") if model_display else Text(),
        Text(f"  {cwd_display}", style="dim"),
        session_display,
    ]

    from iac_code.utils.log import is_debug_enabled

    if is_debug_enabled():
        from iac_code.config import get_config_dir

        log_path = get_config_dir() / "logs" / "latest.log"
        items.append(Text())
        items.append(Text("  {}".format(_("Debug mode")), style="bold yellow"))
        items.append(Text("  {}: {}".format(_("Log file"), log_path), style="dim yellow"))

    return Panel(Group(*items), border_style=ACCENT, expand=True)


def print_welcome_banner(
    console: Console,
    model: str,
    cwd: str,
    session_id: str | None = None,
    session_name: str | None = None,
) -> None:
    """Print the banner with a high-resolution logo when supported."""
    from iac_code.ui.terminal_image import (
        IMAGE_COLUMNS,
        IMAGE_ROWS,
        build_terminal_image_escape,
        detect_terminal_image_protocol,
        load_terminal_logo_png,
        terminal_cell_size,
    )

    protocol = detect_terminal_image_protocol(console)
    image_escape = ""
    if protocol is not None:
        try:
            image_escape = build_terminal_image_escape(
                protocol,
                load_terminal_logo_png(),
                columns=IMAGE_COLUMNS,
                rows=IMAGE_ROWS,
                cell_size=terminal_cell_size(console),
            )
        except (OSError, ValueError):
            protocol = None

    panel = render_welcome_banner(
        model,
        cwd,
        session_id=session_id,
        session_name=session_name,
        _image_placeholder=protocol is not None,
    )
    rendered_height = len(console.render_lines(panel, console.options, pad=False)) if protocol is not None else 0
    console.print(panel)

    if protocol is None or not image_escape:
        return

    # After Console.print(), the cursor is one line below the panel. The image
    # placeholder begins on panel row 4 and after five columns (border,
    # padding, and the three-cell logo indent). Drawing between DECSC/DECRC
    # keeps the prompt position unchanged for Kitty, iTerm2, and Sixel.
    logo_top_row = 4
    cursor_up = max(1, rendered_height - logo_top_row)
    overlay = "\0337\033[{}A\033[5C{}\0338".format(cursor_up, image_escape)
    try:
        console.file.write(overlay)
        console.file.flush()
    except (AttributeError, OSError, UnicodeError):
        # The text banner has already been printed. Output streams that reject
        # control sequences remain usable; the next redraw will use detection
        # again and can fall back to Braille.
        return

File: iac-code-rs/crates/iac-code-tui/src/banner.rs
use std::path::PathBuf;

use crate::width::terminal_display_width;

mod formatting;
mod panel;

use formatting::{capitalize_username, display_cwd, shell_quote};
use panel::{ansi_style, panel_row, AnsiStyle, Segment};

pub const LOGO_LINES: [&str; 5] = [
    "      ▄▄███▄▄      ",
    "   ▄██████████▄▄   ",
    " ▄█▀████████████▄  ",
    "████████████████████",
    " ▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀ ",
];

pub const ACCENT: &str = "bright_cyan";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WelcomeBannerLabels {
    pub welcome_back: String,
    pub description: String,
    pub session: String,
    pub debug_mode: String,
    pub log_file: String,
}

impl Default for WelcomeBannerLabels {
    fn default() -> Self {
        Self {
            welcome_back: "Welcome back".to_owned(),
            description: "Your AI-powered Infrastructure as Code assistant".to_owned(),
            session: "Session".to_owned(),
            debug_mode: "Debug mode".to_owned(),
            log_file: "Log file".to_owned(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct WelcomeBannerState {
    pub model: String,
    pub cwd: PathBuf,
    pub version: String,
    pub username: String,
    pub provider_display: Option<String>,
    pub home_dir: Option<PathBuf>,
    pub session_id: Option<String>,
    pub session_name: Option<String>,
    pub debug_log_path: Option<PathBuf>,
    pub labels: WelcomeBannerLabels,
}

impl WelcomeBannerState {
    pub fn new(
        model: impl Into<String>,
        cwd: impl Into<PathBuf>,
        version: impl Into<String>,
    ) -> Self {
        Self {
            model: model.into(),
            cwd: cwd.into(),
            version: version.into(),
            username: "User".to_owned(),
            provider_display: None,
            home_dir: std::env::var_os("HOME").map(PathBuf::from),
            session_id: None,
            session_name: None,
            debug_log_path: None,
            labels: WelcomeBannerLabels::default(),
        }
    }

    pub fn with_username(mut self, username: impl Into<String>) -> Self {
        let username = username.into();
        self.username = capitalize_username(&username);
        self
    }

    pub fn with_provider_display(mut self, provider_display: impl Into<String>) -> Self {
        let provider_display = provider_display.into();
        if !provider_display.is_empty() {
            self.provider_display = Some(provider_display);
        }
        self
    }

    pub fn with_home_dir(mut self, home_dir: impl Into<PathBuf>) -> Self {
        self.home_dir = Some(home_dir.into());
        self
    }

    pub fn with_session(
        mut self,
        session_id: impl Into<String>,
        session_name: Option<&str>,
    ) -> Self {
        self.session_id = Some(session_id.into());
        self.session_name = session_name.map(str::to_owned);
        self
    }

    pub fn with_debug_log_path(mut self, debug_log_path: impl Into<PathBuf>) -> Self {
        self.debug_log_path = Some(debug_log_path.into());
        self
    }

    pub fn with_labels(mut self, labels: WelcomeBannerLabels) -> Self {
        self.labels = labels;
        self
    }

    pub fn cwd_display(&self) -> String {
        display_cwd(&self.cwd, self.home_dir.as_deref())
    }

    pub fn model_display(&self) -> Option<String> {
        if self.model.is_empty() {
            return None;
        }
        Some(match &self.provider_display {
            Some(provider) if !provider.is_empty() => format!("{provider} / {}", self.model),
            _ => self.model.clone(),
        })
    }

    pub fn session_display(&self) -> Option<String> {
        match (self.session_id.as_deref(), self.session_name.as_deref()) {
            (Some(session_id), Some(session_name)) if !session_name.is_empty() => Some(format!(
                "{}: {session_name} ({session_id})",
                self.labels.session
            )),
            (Some(session_id), _) => Some(format!("{}: {session_id}", self.labels.session)),
            _ => None,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BannerUpdate {
    pub current_version: String,
    pub version: String,
    pub update_command: Vec<String>,
    pub release_notes_url: Option<String>,
}

impl BannerUpdate {
    pub fn new(
        current_version: impl Into<String>,
        version: impl Into<String>,
        update_command: Vec<String>,
    ) -> Self {
        Self {
            current_version: current_version.into(),
            version: version.into(),
            update_command,
            release_notes_url: None,
        }
    }

    pub fn with_release_notes_url(mut self, release_notes_url: impl Into<String>) -> Self {
        self.release_notes_url = Some(release_notes_url.into());
        self
    }
}

pub fn render_welcome_banner_lines(state: &WelcomeBannerState) -> Vec<String> {
    let mut lines = vec![
        String::new(),
        format!("  {} {}!", state.labels.welcome_back, state.username),
        String::new(),
    ];
    lines.extend(LOGO_LINES.iter().map(|line| format!("   {line}")));
    lines.push(format!("  {}", state.labels.description));
    lines.push(String::new());
    lines.push(format!("  iac-code v{}", state.version));
    if let Some(model_display) = state.model_display() {
        lines.push(format!("  {model_display}"));
    }
    lines.push(format!("  {}", state.cwd_display()));
    if let Some(session_display) = state.session_display() {
        lines.push(format!("  {session_display}"));
    }
    if let Some(log_path) = &state.debug_log_path {
        lines.push(String::new());
        lines.push(format!("  {}", state.labels.debug_mode));
        lines.push(format!(
            "  {}: {}",
            state.labels.log_file,
            log_path.display()
        ));
    }
    lines
}

pub fn render_welcome_banner_ansi_lines(
    state: &WelcomeBannerState,
    terminal_width: usize,
) -> Vec<String> {
    let width = terminal_width.max(20);
    let inner_width = width.saturating_sub(2);
    let mut lines = Vec::new();

    lines.push(ansi_style(
        AnsiStyle::Cyan,
        &format!("╭{}╮", "─".repeat(inner_width)),
    ));
    lines.push(panel_row(inner_width, vec![]));
    lines.push(panel_row(
        inner_width,
        vec![
            Segment::plain(" "),
            Segment::styled(
                AnsiStyle::Bold,
                format!("  {} {}!", state.labels.welcome_back, state.username),
            ),
        ],
    ));
    lines.push(panel_row(inner_width, vec![]));

    for (index, line) in LOGO_LINES.iter().enumerate() {
        let logo_text = format!("   {line}");
        let mut segments = vec![
            Segment::plain(" "),
            Segment::styled(AnsiStyle::Cyan, logo_text.clone()),
        ];
        if index == 2 {
            let current_width = 1 + terminal_display_width(&logo_text);
            let description_width = terminal_display_width(&state.labels.description);
            let desired_width = (inner_width * 62 / 100).max(current_width + 5);
            let max_width = inner_width.saturating_sub(description_width);
            let target_width = desired_width.min(max_width).max(current_width + 5);
            segments.push(Segment::plain(
                " ".repeat(target_width.saturating_sub(current_width)),
            ));
            segments.push(Segment::styled(
                AnsiStyle::ItalicWhite,
                state.labels.description.clone(),
            ));
        }
        lines.push(panel_row(inner_width, segments));
    }

    lines.push(panel_row(inner_width, vec![]));
    lines.push(panel_row(
        inner_width,
        vec![
            Segment::plain(" "),
            Segment::styled(AnsiStyle::Dim, format!("  iac-code v{}", state.version)),
        ],
    ));
    if let Some(model_display) = state.model_display() {
        lines.push(panel_row(
            inner_width,
            vec![
                Segment::plain(" "),
                Segment::styled(AnsiStyle::Dim, format!("  {model_display}")),
            ],
        ));
    }
    lines.push(panel_row(
        inner_width,
        vec![
            Segment::plain(" "),
            Segment::styled(AnsiStyle::Dim, format!("  {}", state.cwd_display())),
        ],
    ));
    if let Some(session_display) = state.session_display() {
        lines.push(panel_row(
            inner_width,
            vec![
                Segment::plain(" "),
                Segment::styled(AnsiStyle::Dim, format!("  {session_display}")),
            ],
        ));
    }

    if let Some(log_path) = &state.debug_log_path {
        lines.push(panel_row(inner_width, vec![]));
        lines.push(panel_row(
            inner_width,
            vec![
                Segment::plain(" "),
                Segment::styled(
                    AnsiStyle::BoldYellow,
                    format!("  {}", state.labels.debug_mode),
                ),
            ],
        ));
        lines.push(panel_row(
            inner_width,
            vec![
                Segment::plain(" "),
                Segment::styled(
                    AnsiStyle::DimYellow,
                    format!("  {}: {}", state.labels.log_file, log_path.display()),
                ),
            ],
        ));
    }

    lines.push(ansi_style(
        AnsiStyle::Cyan,
        &format!("╰{}╯", "─".repeat(inner_width)),
    ));
    lines
}

pub fn render_update_prompt_header_lines(update: &BannerUpdate) -> Vec<String> {
    let mut lines = vec![
        format!(
            "Update available! {} -> {}",
            update.current_version, update.version
        ),
        format!(
            "Update command: {}",
            format_update_command(&update.update_command)
        ),
    ];
    if let Some(url) = &update.release_notes_url {
        lines.push(format!("Release notes: {url}"));
    }
    lines
}

pub fn render_update_notice_lines(update: &BannerUpdate) -> Vec<String> {
    let mut lines = vec![
        format!(
            "Update available! {} -> {}",
            update.current_version, update.version
        ),
        format!(
            "Run {} to update.",
            format_update_command(&update.update_command)
        ),
    ];
    if let Some(url) = &update.release_notes_url {
        lines.push(format!("Release notes: {url}"));
    }
    lines
}

pub fn format_update_command(command: &[String]) -> String {
    command
        .iter()
        .map(|part| shell_quote(part))
        .collect::<Vec<_>>()
        .join(" ")
}
